feat: Allow segment-specific styling

heyrict opened the pull request:
#### TODOs
- [x] segment-specific configuration should accept both string (value) and table of { value, style }.
- [ ] prefix or suffix should accpet both `false` (to disable) and string (value)
- [x] ~~update docs (since all segments should be properly documented)~~ Reverted. Will leave this undocumented for a while.

#### Description
This PR will extend current configuration file as described in [#132(comment)](https://github.com/starship/starship/issues/132#issuecomment-520172694). Making it work won't have any breaking changes and is easy to implement.

Character-specific styling will possibly have some breaking changes and will not be included in this PR.

#### Motivation and Context
Related [#132(comment)](https://github.com/starship/starship/issues/132#issuecomment-520172694)

#### Screenshots
Rendered with the following config.
```toml
[git_status]
prefix = ""
suffix = " "

	[git_status.untracked]
	value = "•"
	style = "blue"

	[git_status.modified]
	value = "•"
	style = "red"
```

![Screenshot from 2019-09-16 10-56-24](https://user-images.githubusercontent.com/25698503/64932323-b2d8cc00-d870-11e9-8866-03bf9bac08bf.png)

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

New feature or breaking change? Previous configuration should work without change at present.

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [ ] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:

- [x] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

heyrict commented:
Not all modules are fully tested. `Segments` in the docs are found using `grep new_segment src/modules/*.rs`.

> prefix or suffix should accpet both false (to disable) and string (value)

I'll leave this unimplemented at present as it is deprecated since we can use an empty string.

chipbuster commented:
@heyrict I went ahead and changed the title of this PR since we had two PRs right next to each other that both were about refactoring the config.

bijancn commented on docs/config/README.md:
````diff
@@ -40,6 +40,28 @@ are segments within it. Every module also has a prefix and suffix that are the d
+style = "red"
+```
+
+Another is segment-wide styling. Segments in the configuration can be a string, or a table with optional `style` and `value` key. In the following config, starship will print `ahead` segment green, `diverged` segment yellow, and other segments red. **Note: Segment-wide stylings will always overwrite module-wide stylings, no matter if the module use conditional stylings such as [character](#charac ...
````
I think it would be `stylings such as the [character]`

bijancn commented on src/config.rs:
```diff
@@ -166,6 +167,38 @@ impl Config for Table {
+    /// Returns `Some(SegmentConfig)` if key exists in the configuration, else `None`.
+    fn get_as_segment_config(&self, key: &str) -> Option<SegmentConfig> {
+        let segment_config = self.get_config(key)?;
+        match segment_config {
```
how about
```
self.get_config(key).map(|segmentconfig| {
  match segment_config ...
})
```

heyrict replied:
Reasonable.

heyrict commented on docs/config/README.md:
```diff
@@ -213,15 +240,20 @@ can do this in two ways: by changing color (red/green) or by changing its shape
+| `style_failure` | `"bold red"`   | The style used if the last command failed.         |
+| `disabled`      | `false`        | Disables the `character` module.                   |
+
+### Segments
```
What about using `Segment("<Symbol>", "style")` as the default value in the table and link it to the `Styling` Section?

chipbuster replied:
Wouldn't we still need to include the segment names somewhere?

chipbuster replied:
My concern is that if you're not used to thinking about the modularity of starship, it's not clear why we even need the segment names (and I think it's a little much to expect someone to read the parts on styling before looking for a setting).

So the segment section is valuable to someone who knows a lot about starship and wants to configure it in detail, but if you're a new user, it's just confusing as to what it is and why there's so much space dedicated to it.

heyrict commented on docs/config/README.md:
````diff
@@ -40,6 +40,28 @@ are segments within it. Every module also has a prefix and suffix that are the d
  "via "         "⬢"        "v10.4.1"       ""
 ```
 
+### Styling
````
Moving to the advanced guide sounds reasonable to me, but we have to provide some info about which config can be configured with a `{value, style}` table.
Any ideas about how to do this in a cleaner way?

heyrict commented:
@chipbuster I don't know if it is the best way to handle this, but I've

- Moved the `styling` section to advanced config.
- Added links to the `styling` section to the default value in the table, instead of using a separate `Segments` section.

Instead of getting stuck with the docs, leaving the docs unchanged may be an option worth considering in the current state. (There will be further changes on segment specific styling in #383)

chipbuster commented:
@heyrict Good idea! We can leave them undocumented for now and then add docs as a TODO item once #383 gets merged as well.

chipbuster commented:
Well, our Java CI is broken so we're stuck waiting for that...

@heyrict If I don't notice before you, go ahead and ping me once our CI is working again so we can get this tested and merged!

heyrict commented:
@chipbuster Ready for merge.

matchai commented:
Hey @jasongaare,

The API has since changed [back in v0.45.0](https://starship.rs/migrating-to-0.45.0/), so I'd suggest referring to the up-to-date [Git Status docs](https://starship.rs/config/#git-status).

For the example above, what used to be:
```toml
[git_status]
prefix = ""
suffix = " "

	[git_status.untracked]
	value = "•"
	style = "blue"

	[git_status.modified]
	value = "•"
	style = "red"
```

Will now be the following:
```toml
[git_status]
untracked = "[•](blue)"
modified = "[•](red)
```

jasongaare commented:
Thanks so much for response! I appreciate you confirming I'm at least _attempting_ to do it correctly. 

This is still not working for me: 

```toml
[git_status]
untracked = "!${count}(blue)"
...
```

so, there must be something else going on with my config (aka user-error!) 
I'll get to the bottom of it, thanks again

jasongaare commented:
In case anyone stumbles upon this, my issue was not first wrapping the format I wanted colored in`[ ]` 

```toml
[git_status]
untracked = "[!${count}](blue)" # works!
```
